src/gc-common.c

```diff
@@ -478,6 +480,45 @@ JL_DLLEXPORT void jl_finalize(jl_value_t *o)
     jl_finalize_th(jl_current_task, o);
 }
 
+int erase_finalizer_at(arraylist_t *list, jl_value_t *o, size_t idx)
+{
+    void **items = list->items;
+    void *v = items[idx];
+    if (o == (jl_value_t*)gc_ptr_clear_tag(v, 1)) {
+        for (size_t j = idx + 2; j < list->len; j += 2) {
+            items[j-2] = items[j];
+            items[j-1] = items[j+1];
+        }
+        list->len = list->len - 2;
+        return 1;
+    }
+    return 0;
+}
+
+int erase_finalizer(arraylist_t *list, jl_value_t *o)
+{
+    for (size_t i = 0; i < list->len; i += 2) {
+        if (erase_finalizer_at(list, o, i))
+            return 1;
+    }
+    return 0;
+}
+
+// Remove the finalizer for `o` from `ct->ptls->finalizers` and cancel that finalizer.
+// `lookup_idx` is the index at which the finalizer was registered in `ct->ptls->finalizers`
```

Comment:
This will just be a random pointer though. After the next GC we will have no idea where the finalizer index will be without scanning for it.

Reply:
Yes. If a GC occurs between the finalizer registration and the call to `Core._cancel_finalizer`, this index becomes meaningless, but in that case, the slow path of `jl_cancel_finalizer` will handle it by performing a full scan. `lookup_idx` is only used for the fast path.

Reply:
But doesn't that means we expect this to normally hit the slow case, except when microbenchmarking?

Reply:
no. There are plenty of benchmarks (e.g. BigInt) where we expect many objects to die before GC has run.